casper/src/test/scala/coop/rchain/casper/batch1/MultiParentCasperCommunicationSpec.scala

```diff
@@ -90,88 +70,96 @@ class MultiParentCasperCommunicationSpec extends FlatSpec with Matchers with Ins
-//          deployData10 <- ConstructDeploy.basicDeployData[Effect](10)
-//          block11      <- nodes(0).addBlock(deployData10)
-//
-//          // Cycle of requesting and passing blocks until block #3 from nodes(0) to nodes(1)
-//          _ <- (0 to 8).toList.traverse_[Effect, Unit] { i =>
-//                nodes(1).receive() >> nodes(0).receive() >> nodes(1).receive() >> nodes(0)
-//                  .receive() >> nodes(1).receive() >> nodes(0).receive()
-//              }
-//
-//          // We simulate a network failure here by not allowing block #2 to get passed to nodes(1)
-//
-//          // And then we assume fetchDependencies eventually gets called
-//          _ <- nodes(1).casperEff.fetchDependencies
-//          _ <- nodes(0).receive()
-//          _ <- nodes(0).receive()
-//
-//          _ = nodes(1).logEff.infos.count(_ startsWith "Requested missing block") should be(10)
-//          result = nodes(0).logEff.infos.count(
-//            s => (s startsWith "Received request for block") && (s endsWith "Response sent.")
-//          ) should be(10)
-//          // Oh yes - 10!
-//        } yield result
-//      }
-//  }
+  // TODO this test fails because `TestNode.propagate(nodes)` returns prematurely.
+  it should "ask peers for blocks it is missing and add them" ignore effectTest {
+    def makeDeploy(i: Int): Effect[Signed[DeployData]] =
+      ConstructDeploy.sourceDeployNowF(
+        Vector("@2!(2)", "@1!(1)")(i),
+        sec = if (i == 0) ConstructDeploy.defaultSec else ConstructDeploy.defaultSec2
+      )
+
+    def stepSplit(nodes: Seq[TestNode[Effect]]) =
+      for {
+        _ <- makeDeploy(0) >>= (nodes(0).addBlock(_))
+        _ <- makeDeploy(1) >>= (nodes(1).addBlock(_))
+
+        _ <- nodes(0).syncWith(nodes(1))
+        _ <- nodes(2).shutoff() //nodes(2) misses this block
+      } yield ()
+
+    def stepSingle(nodes: Seq[TestNode[Effect]]) =
+      for {
+        _ <- makeDeploy(0) >>= (nodes(0).addBlock(_))
+
+        _ <- nodes(1).syncWith(nodes(0))
+        _ <- nodes(2).shutoff() //nodes(2) misses this block
+      } yield ()
+
+    TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
+      for {
+        _ <- stepSplit(nodes) // blocks a1 a2
+        _ <- stepSplit(nodes) // blocks b1 b2
+        _ <- stepSplit(nodes) // blocks c1 c2
+
+        _ <- stepSingle(nodes) // block d1
+        _ <- stepSingle(nodes) // block e1
+
+        _ <- stepSplit(nodes) // blocks f1 f2
+        _ <- stepSplit(nodes) // blocks g1 g2
+
+        // this block will be propagated to all nodes and force nodes(2) to ask for missing blocks.
+        br <- makeDeploy(0) >>= (nodes(0).addBlock(_)) // block h1
+
+        _ <- TestNode.propagate(nodes) // force the network to communicate
+
+        _ <- nodes(2).contains(br.blockHash) shouldBeF true
+
+        nr <- makeDeploy(0) >>= (nodes(2).addBlock(_))
+      } yield {
+        PrettyPrinter.buildString(nr.header.parentsHashList) shouldBe PrettyPrinter
```

Comment:
`buildString` would truncate the whole message which can be a problem in some extreme cases. Maybe `buildStringNoLimit` or just compare the hashes

Reply:
Nice catch, I agree. I don't like comparing just hashes because in case of test failure it does not print valuable information - what the hashes are actually, only non readable byte string.

Reply:
https://github.com/rchain/rchain/pull/3355/commits/fe99286a4e6aade733a95e754960f688475dc21e

Reply:
What is printed is whatever is returned on calling `.toString` on that object. We can create override of `toString` to customize  friendly string representation which is also useful in debugging.
I'm not sure if ScalaTest have some other way to do it e.g. using cats with Show instances which would be more elegant solution.